protocol/lavaprotocol/finalization_consensus.go

```diff
@@ -100,83 +99,113 @@ func (fc *FinalizationConsensus) insertProviderToConsensus(blockDistanceForFinal
+		if foundDiscrepancy {
+			blockToHashesToAgreeingProviders = fc.prevEpochBlockToHashesToAgreeingProviders
+		} else {
+			// No discrepancy found, log and return nil
+			logSuccessUpdate()
+			return finalizationConflict, nil
 		}
+	}
+
+	replyFinalization := conflicttypes.NewRelayFinalizationMetaDataFromRelaySessionAndRelayReply(relaySession, reply, consumerAddress)
+	finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: &replyFinalization}
 
-		// check for discrepancy with old epoch
-		for idx, consensus := range fc.prevEpochProviderHashesConsensus {
-			err := fc.discrepancyChecker(finalizedBlocks, consensus)
-			if err != nil {
-				// TODO: bring the other data as proof
-				finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: reply}
-				return finalizationConflict, utils.LavaFormatError("Simulation: prev epoch Conflict found in discrepancyChecker", err, utils.Attribute{Key: "Consensus idx", Value: strconv.Itoa(idx)}, utils.Attribute{Key: "provider", Value: providerAddress})
+	var otherBlockHash string
+	for blockHash, agreeingProviders := range blockToHashesToAgreeingProviders[discrepancyBlock] {
+		if providerPrevReply, ok := agreeingProviders[providerAddress]; ok {
+			// Same provider conflict found - Report and block provider
+			relayFinalization, errWrapped := fc.createRelayFinalizationFromProviderDataContainer(providerPrevReply, consumerAddress)
+			if errWrapped != nil {
+				return nil, errWrapped
 			}
+
+			logSuccessUpdate()
+			finalizationConflict.RelayReply1 = relayFinalization
+
+			errWrapped = sdkerrors.Wrap(lavasession.BlockProviderError, fmt.Sprintf("found same provider conflict on block [%d], provider address [%s]", discrepancyBlock, providerAddress))
+			return finalizationConflict, errWrapped
+		} else if otherBlockHash == "" {
+			// Use some other hash to create proof, in case of no same provider conflict
+			otherBlockHash = blockHash
 		}
 	}
-	if debug {
-		utils.LavaFormatDebug("finalization information update successfully", utils.Attribute{Key: "specId", Value: fc.specId}, utils.Attribute{Key: "finalization data", Value: finalizedBlocks}, utils.Attribute{Key: "currentProviderHashesConsensus", Value: fc.currentProviderHashesConsensus}, utils.Attribute{Key: "currentProviderHashesConsensus", Value: fc.currentProviderHashesConsensus})
+
+	var otherProviderAddress string
+	// Same provider conflict not found, create proof with other provider
+	for providerAddress, providerDataContainer := range blockToHashesToAgreeingProviders[discrepancyBlock][otherBlockHash] {
+		// Finalization conflict
+		relayFinalization, err := fc.createRelayFinalizationFromProviderDataContainer(providerDataContainer, consumerAddress)
+		if err != nil {
+			return nil, err
+		}
+
+		finalizationConflict.RelayReply1 = relayFinalization
+		otherProviderAddress = providerAddress
+		// TODO: Should we send a conflict proof for all providers?
+		break
```

Comment:
This is a special case, where I only need one disagreeing provider, and that's it. That's why I break on first provider, and that's why I don't need to sort it. I don't want it to be deterministic (it's actually better this way, since this way may "catch" more lying providers)

Reply:
I went through the code again and this non-determinism is problematic. If you follow the code, the `finalizationConflict` object is eventually an input to the consensus code of the Detection TX. Currently, there is no real handling of two providers conflict (other than printing an event) so it's harmless now. But when this mechanism will be implemented in the future, it's very likely that the `finalizationConflict` will be written to the state so it has to be deterministic.